src/AudioIO.cpp

```diff
@@ -1462,8 +1462,9 @@ void AudioIO::StopStream()
 
       while( mAudioThreadShouldCallTrackBufferExchangeOnce )
       {
-         // LLL:  Experienced recursive yield here...once.
-         wxTheApp->Yield(true); // Pass true for onlyIfNeeded to avoid recursive call error.
+         //FIXME: Seems like this block of the UI thread isn't bounded,
+         //but we cannot allow event handlers to see incompletely terminated
+         //AudioIO state with wxYield (or similar functions)
```

Comment:
Study the function `AudioThread::ExitCode AudioThread::Entry()` for the thread that must make the change of state that un-blocks this main thread.  @crsib has become more familiar with this thread, after reviewing other PRs of mine that refactor the functions called in it.

I strongly doubt that this other thread ever fails to make progress now, or to change the condition that the main thread is waiting on here.  If it ever did, then the loop above would have effectively become the new alternative event-loop of Audacity, making the program still responsive to some inputs, and yet the program would be stuck in a strange state, unable to stop a play or recording and start another one.  But I recollect no such bugs.

The Yield() call does nothing to assist this other thread in making progress, which is why I favor removing this ancient and unexplained piece of code.

Reply:
And, I am very aware that this bit of inter-thread communication with a volatile bool variable is not correct.  Leave it to me to rewrite it better, as part of my sweep of all the bad old stuff in the audio engine that I want to complete in the coming release.

This looks like the sort of thing that should be done the modern way with a std::condition_variable, and a companion std::mutex.  And it would not be bad to do so, because the other thread is not the low-latency thread.  We don't need extra effort to avoid memory allocations or mutex locking in it.

Reply:
The Yield() seems only to have been a source of bugs, because it causes some event callbacks to be invoked while still in the middle of other, un-finished event callbacks, causing surprising things to happen because the callbacks were not written in anticipation of this.

The Yield() call is NOT a yielding of the thread to other threads (like `std::this_thread::yield()`) but rather allows more event dispatches to happen while still in the main thread, as a sort of cooperative multi-tasking.

The original author of this Yield() thread (some time before the 2010 event horizon of this git repository) might just have been confused, and nobody has yet paused long enough to recognize that confusion.  Or maybe they really were trying to compensate for bugs in ancient versions of AudioThread, but these bugs were fixed in better ways by other means.

Reply:
And you have my great respect for working backwards from a crashing call stack to this Yield() as the ulterior cause of it.  It is very non-obvious.

Reply:
Thank you!

Reply:
I agree that the code in StopStream is quite broken, and the check there doesn't work.

What bothers me here - is there a chance that the app will hang in the new loop? The previous version effectively created a new "run loop," which is also wrong but more user-friendly.

Should we think of a better synchronization here now? Like using atomics for mStreamToken and waiting on a proper condition before we exit. I can easily imagine a case where `mAudioThreadShouldCallTrackBufferExchangeOnce` is true till the end of the world. Should we just wait on mStreamToken?

Reply:
We've debugged with Vitaly, probably it wont hang :)

Reply:
Sure, this needs rewriting.  In practice, the Audio thread doesn't take long to finish its work.

We could use a std condition variable and wait_for if we are worried about that contingency, but I don't know how the main program could recover if the Audio thread is not making progress.  Kill the thread?  Leaving things in what undefined state?  I think there would be nothing to do but put up an error dialog, then quit the program after the acknowledgment.

But it is not a known problem in my experience.

I'm speculating that the ancient code yielding to the event dispatcher MIGHT have helped twenty years ago, with slower drives recording the samples, to allow refreshes of windows to happen, in case the delay of the other thread was noticeable.  Maybe it prevented visual ugliness.

The problem is yeilding to all events indiscriminately.

Reply:
After even more discussion with Vitaly, I think we should probably just use `std::this_thread::yield();` instead of both of the instructions.

Reply:
But this code is fine as well!

Reply:
So you think a busy-wait, but with a yield, would be preferable to the repeated sleeps?  I have had that thought too.